enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,272 @@
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@spadgett, console team"
+  - "@wking, upgrades team"
+  - "@sdodson, upgrades team"
+  - "@LalatenduMohanty, upgrades team"
+  - "@soltysh, workloads team"
+approvers:
+  - TBD
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - None
+creation-date: 2022-07-29
+last-updated: 2022-07-29
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/USHIFT-233
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1850656
+---
+
+# Advertise Version Details
+
+## Summary
+
+We have several use cases for allowing authenticated users to discover
+whether they are communicating with OCP or MicroShift, including the
+version of the product. We already have the ClusterVersion resource
+for OCP, but that requires special privileges to read and is not
+present in a MicroShift deployment. This enhancement describes a
+standard that both platforms can use to share platform and version
+information with API callers.
+
+## Motivation
+
+Users configuring applications to deploy on different versions of
+OpenShift may need to adjust their configuration to work differently
+on OCP and MicroShift. Ideally this is not needed, but in practice it
+is useful because of different APIs being available and different
+constraints being in place. It is possible for a cluster admin to read
+the ClusterVersion resource in an OCP cluster to discover that it is
+OCP and what version is running. This approach has two
+drawbacks. First, not all console users have cluster admin privileges,
+which means that the console cannot show them the cluster version
+details. Second, the ClusterVersion API is not present in a MicroShift
+deployment, so instead [we created a
+ConfigMap](https://github.com/openshift/microshift/pull/776) in the
+`openshift-public` namespace containing some basic version
+details. That means a caller must know two different APIs to check the
+version and product they are communicating with. This enhancement
+describes how we can standardize on one to make it easier for all
+authenticated users to obtain basic version details for a cluster.
+
+### User Stories
+
+As a console user, I want to see the version of OCP I am using so that
+I can refer to the correct version of the documentation and API
+references.
+
+As an authenticated API caller, I want to obtain the product name and
+version of the OpenShift variant I am communicating with so that I can
+use it to make decisions on the client side.
+
+### Goals
+
+* Avoid introducing a new CRD, so we do not increase the overhead for
+  MicroShift.
```

Comment:
how big a concern is this?  

not that a configmap is terrible for something like this, but i think we should generally be preferring validate-able, discoverable apis.

Reply:
Every megabyte counts at this point.

Reply:
does a CRD represent a megabyte?  (i'm just genuinely curious at this point what the cost of an additional CRD api is)

Reply:
I tend to find the simplicity argument more compelling, but I included this for completeness. I can sketch out an implementation if you think having the numbers would be useful.

Reply:
i don't know if it's worth the effort to determine the size, especially if you're committed to this path for the other reasons anyway, but if we don't have any meaningful data that shows just how much memory this would cost us, it feels suspect to list it as a reason here.  Was there at least someone from the apiserver team or elsewhere who provided a rough estimate of the associated cost?

Reply:
Adding a module with a simple struct and the asset with the CRD increases the binary size about 20K to the static binary size, which affects RAM and over-the-wire update size. I didn't measure the RAM use to see if it was significantly more than that 20K.

Reply:
i'm going to mark this resolved, seems like we're staying on the configmap path and while i have reservations about that, they aren't worth holding up this discussion.